### ISSUE TITLE
Metrics remote-write falling behind when reading checkpoint

### ISSUE THREAD

**Issue by bboreham** [grafana/agent#701]
[I suspect this applies to Prometheus remote_write, but the stack traces go through the Grafana fork so filing here]

I observed periodic peaks in send lag (`prometheus_remote_storage_highest_timestamp_in_seconds - prometheus_remote_storage_queue_highest_sent_timestamp_seconds`:

![image](https://user-images.githubusercontent.com/8125524/123937391-048f0f80-d98e-11eb-9ce0-c2804ab194bb.png)

The timing coincides with creating a checkpoint.

After studing many goroutine dumps, I formed the opinion that `Watcher.readCheckpoint()` can hold the `QueueManager.seriesMtx` lock for extended periods:

```
goroutine 60654572 [runnable]:
github.com/prometheus/prometheus/pkg/intern.(*pool).Release(0xc000353350, 0xc002038360, 0x8)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20210608193638-7b78de4ccffc/pkg/intern/intern.go:126 +0x214
github.com/prometheus/prometheus/pkg/intern.ReleaseLabels(0x426e880, 0xc000353350, 0xc9167d7a40, 0xe, 0xe)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20210608193638-7b78de4ccffc/pkg/intern/intern.go:163 +0x89
github.com/prometheus/prometheus/storage/remote.(*QueueManager).StoreSeries(0xc000602d80, 0xc2eaf18000, 0xe2, 0x100, 0x324a)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20210608193638-7b78de4ccffc/storage/remote/queue_manager.go:656 +0x36a
github.com/prometheus/prometheus/tsdb/wal.(*Watcher).readSegment(0xc0003a4540, 0xc02e97a000, 0x324a, 0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20210608193638-7b78de4ccffc/tsdb/wal/watcher.go:486 +0x744
github.com/prometheus/prometheus/tsdb/wal.(*Watcher).readCheckpoint(0xc0003a4540, 0xc142cd64e0, 0x54, 0x0, 0x0)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20210608193638-7b78de4ccffc/tsdb/wal/watcher.go:586 +0x41f
github.com/prometheus/prometheus/tsdb/wal.(*Watcher).garbageCollectSeries(0xc0003a4540, 0x327e, 0x100000001, 0x51b1ee)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20210608193638-7b78de4ccffc/tsdb/wal/watcher.go:458 +0x5a5
github.com/prometheus/prometheus/tsdb/wal.(*Watcher).watch.func1(0xc8e0d71740, 0xc0003a4540, 0x327e)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20210608193638-7b78de4ccffc/tsdb/wal/watcher.go:376 +0x85
created by github.com/prometheus/prometheus/tsdb/wal.(*Watcher).watch
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20210608193638-7b78de4ccffc/tsdb/wal/watcher.go:372 +0x425
```

Holding the lock prevents `QueueManager.Append()` from running so no new data is added to outgoing queues.

As I understand it, `StoreSeries()` will be called with all series from one target, so potentially hundreds of thousands.  Perhaps it should yield the lock every 100 series, something like that?

**Comment by rfratto** [comment]
>  Perhaps it should yield the lock every 100 series, something like that?
Hmm, I think this makes sense to me. I think you might be able to raise some concern that it opens the possibility of a sample being sent without the series labels being stored, but in practice this shouldn't happen: samples are only read concurrently with series when reading from a checkpoint, and the series will always be from the checkpoint where presumably we already have stored the labels for every sample being read. 
 
cc @cstyan

**Comment by rfratto** [comment]
This is resolved upstream and dependent on updating the Prometheus dependency now. We're trying to avoid significant changes until we release v0.19.0 of the agent, but will update the dep following that.

**Comment by rfratto** [comment]
This has been released in v0.21.0.
